File: src/tut01_nodes_py/tut01_nodes_py/listener.py
# Copyright 2026 Jose Laruta
"""Print every message received on the `chatter` topic."""

import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from std_msgs.msg import String


class Listener(Node):
    """Subscribes to `chatter` and logs each message."""

    def __init__(self) -> None:
        super().__init__('listener')
        self._sub = self.create_subscription(String, 'chatter', self._on_message, 10)
        self.get_logger().info('listener started')

    def _on_message(self, msg: String) -> None:
        """Log one received message."""
        self.get_logger().info(f'I heard: "{msg.data}"')


def main(args: list[str] | None = None) -> None:
    """Entry point: init, spin, and shut down cleanly."""
    rclpy.init(args=args)
    node = Listener()
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == '__main__':
    main()

File: src/tut01_nodes/src/listener.cpp
// Copyright 2026 Jose Laruta

#include <functional>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

/// Subscribes to `chatter` and logs each message.
class Listener : public rclcpp::Node
{
public:
  Listener()
  : Node("listener")
  {
    // std::bind is the older way to turn a member function into a callback.
    // It appears once in these lessons, here, because most ROS 2 code you will
    // read uses it. Everywhere else a lambda does the same job.
    sub_ = create_subscription<std_msgs::msg::String>(
      "chatter", 10, std::bind(&Listener::on_message, this, std::placeholders::_1));
    RCLCPP_INFO(get_logger(), "listener started");
  }

private:
  /// Log one received message.
  void on_message(const std_msgs::msg::String & msg)
  {
    RCLCPP_INFO(get_logger(), "I heard: \"%s\"", msg.data.c_str());
  }

  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<Listener>());
  rclcpp::shutdown();
  return 0;
}
